### ISSUE TITLE
Simple pipe crashes after initialising starship

### ISSUE THREAD

**Issue by agoose77** [oilshell/oil#1822]
1. If you use nix, `nix shell nixpkgs#oil nixpkgs#starship`
2. Launch `osh` (0.20.0)
3. Source `starship` with `eval "$(starship init bash)"`
4. Crash with `echo foo | head`

**Comment by andychu** [comment]
Thanks for the report!  What's the output on the terminal?  Does it just say "seg fault" or is there more info?

(BTW people have run starship with OSH before, I believe it should work)

**Comment by jameschensmith** [comment]
Here's the output I got from running the steps provided by OP:

```
% nix shell nixpkgs#oil nixpkgs#starship
% osh --version
Oils 0.20.0             https://www.oilshell.org/

git commit = ffbe52abca49a72a984df0340646e38a597da64f
% starship --version
starship 1.17.1
branch:
commit_hash:
build_time:1980-01-01 00:00:00 +00:00
build_env:rustc 1.75.0 (82e1608df 2023-12-21) (built from a source tarball),
% osh
osh-0.20.0$ echo foo | head
foo
osh-0.20.0$ eval "$(starship init bash)"

~
❯ echo foo | head
osh: parent failed to set process group for PID 42996 to 42994: Operation not permitted
osh: child failed to set process group for PID 42996 to 42994: Operation not permitted
foo
```

**Comment by agoose77** [comment]
Thanks! I updated my unstable, and with osh 0.20 I see the same error rather than a segfault.

**Comment by Melkor333** [comment]
I tried to debug this a bit over [on zulip](https://oilshell.zulipchat.com/#narrow/stream/264891-oil-help/topic/history.20.7C.20less.20broken.3F).
TLDR:
> I think the issue is ordering
> Bash: starship_preexec, head, echo
> Osh: echo, starship_preexec, head

I think what this needs is either [here](https://github.com/oilshell/oil/blob/master/osh/cmd_eval.py#L834) or [here](https://github.com/oilshell/oil/blob/1d07844f366158ce5bc3e1db34670138a735d0fb/osh/cmd_eval.py#L2046) a conditional to check if there is a preceding `|` (or following? it confuses me that the `debug` is run after the `echo` command).

**Comment by andychu** [comment]
OK thank you, I've reproduced this bug!

With the Nix build I get the "operation not permitted" errors

I'm also getting ths with the dev build, hm

```
❯ echo hi|head
oils I/O error (main): No such process
```

**Comment by andychu** [comment]
This is fixed, and will be out with the next release.  Thanks for testing!

This is a hairy interaction between three features: `shopt -s lastpipe`, POSIX job control, and  the DEBUG trap

Starship uses the third feature, and the first 2 are in the interactive shell.  Added this quirk to the docs:

http://travis-ci.oilshell.org/github-jobs/6428/ovm-tarball.wwz/_release/VERSION/doc/quirks.html#with-job-control-the-debug-trap-is-disabled-for-the-last-part-of-a-pipeline

**Comment by andychu** [comment]
Released!  https://www.oilshell.org/blog/2024/03/release-0.21.0.html
